### ISSUE TITLE
Cannot override headerStyle in nested navigator

### ISSUE THREAD

**Issue by chochihim** [react-navigation/react-navigation#2580]
I have nested navigators as follow:

```js
const MyScreen = () => <View />;
MyScreen.navigationOptions = {
  headerStyle: {
    backgroundColor: 'yellow'
  }
}

const MyTabNavigator = TabNavigator(
  {
    screen1: {
        screen: MyScreen
    }
  },
  {
    navigationOptions: {
      headerStyle: {
        backgroundColor: 'red'
      }
    }
  }
)

const MyStackNavigator = StackNavigator(
  {
    route1: {
        screen: MyTabNavigator
    }
  },
  {
    navigationOptions: {
      headerStyle: {
        backgroundColor: 'blue'
      }
    }
  }
)
```

I think screen's `navigationOptions` should override navigator's but it does not. In my example, the header background color is blue (i.e. StackNavigator takes effect). If I remove `navigationOptions` from `MyStackNavigator`, the color is yellow which works as expected (screen options overriding navigators).

| software         | version
| ---------------- | -------
| react-navigation | 1.0.0-beta.11
| react-native     |  0.47.2
| node             | v6.9.1
| npm or yarn      | npm 3.10.8

**Comment by matthamil** [comment]
Does this still occur when you set `navigationOptions` as a function instead of an object? See https://reactnavigation.org/docs/navigators/navigation-options

**Comment by chochihim** [comment]
@matthamil yes it still occur:

```js
const MyScreen = () => <View />;
MyScreen.navigationOptions = () => ({
    headerStyle: {
        backgroundColor: 'yellow'
    }
})

const MyTabNavigator = TabNavigator(
    {
        screen1: {
            screen: MyScreen
        }
    },
    {
        navigationOptions: () => ({
            headerStyle: {
                backgroundColor: 'red'
            }
        })
    }
)

const MyStackNavigator = StackNavigator(
    {
        route1: {
            screen: MyTabNavigator
        }
    },
    {
        navigationOptions: () => ({
            headerStyle: {
                backgroundColor: 'blue'
            }
        })
    }
);
```

**Comment by Flavien** [comment]
I agree it would make sense to be able to set it on the nested element.

**Comment by Flavien** [comment]
A workaround is to do:

```
MyTabNavigator.navigationOptions = {
            headerStyle: {
                backgroundColor: 'red'
            }
        };

**Comment by kelset** [comment]
Pinging OP @chochihim since this issue has not been active for a while, and it's related to an old version of the lib.

Please let me know if you want this to remain open (does it still happen in the latest version?); if I get no answer in the next 7 days I will close it.

**Comment by chochihim** [comment]
@kelset Sorry for late reply. I just tested in 1.0.0-beta.21 with my original code snippet, and yes, the issue still happens.

I try to use workaround from @Flavien as follow:
```javascript
const MyScreen = () => <View />;
MyScreen.navigationOptions = {
  headerStyle: {
    backgroundColor: 'yellow'
  }
}

const MyTabNavigator = TabNavigator(
  {
    screen1: {
        screen: MyScreen
    }
  }
)

MyTabNavigator.navigationOptions = {
    headerStyle: {
        backgroundColor: 'red'
    }
}

const MyStackNavigator = StackNavigator(
  {
    route1: {
        screen: MyTabNavigator
    }
  },
  {
    navigationOptions: {
      headerStyle: {
        backgroundColor: 'blue'
      }
    }
  }
)

export default MyStackNavigator;
```

And the header is red. That means `headerStyle` defined on `MyScreen` still cannot take effect.

**Comment by satya164** [comment]
`headerStyle` is an option in `StackNavigator`. It won't work when set in `TabNavigator`. The solution @Flavien suggested is the correct way to set the navigation options for the parent navigator.

**Comment by llamaluvr** [comment]
The OP was trying to set the `headerStyle` of the child of a `TabNavigator` and could not until removing the `headerStyle` from the `StackNavigator` one level above the `TabNavigator`. Unless I'm missing something, @Flavien's workaround only moves the `navigationOptions` out of the `TabNavigator` constructor. Even if that works, it's still a bug if `navigationOptions` behaves one way if set in the constructor and another way otherwise.

I also don't think it's correct to stay that `headerStyle` will not work when set in `TabNavigator`, as @satya164 suggested. It does work. In fact, I rely on it to work to get around this bug. Currently, I have a `TabNavigator` where different children need different headers, and I achieve that by writing (very ugly) code in the `TabNavigator`'s `navigationOptions` to examine the current route and determine which style to apply. I would really like to set it on each screen so I wouldn't have to tear apart navigation state just to figure out which style to apply.

The most intuitive behavior is that, if something has a header, you should be able to set the `headerStyle` on it, and, if it doesn't have a `headerStyle`, it inherits the `headerStyle` from its parent. This works with other header properties, which emphasizes why the current behavior is not intuitive.

**Comment by satya164** [comment]
> The OP was trying to set the headerStyle of the child of a TabNavigator and could not until removing the headerStyle from the StackNavigator one level above the TabNavigator.

The OP specified `headerStyle` in `navigationOptions` of `TabNavigator`, and `TabNavigator` doesn't support `headerStyle` option. It's expected behaviour.

> Even if that works, it's still a bug if navigationOptions behaves one way if set in the constructor and another way otherwise.

I believe there is a fundamental misunderstanding here. One is passing options as an argument to the `TabNavigator` function (`const MyScreen = createThisComponent(someScreens, someOptions)`) and the other one is adding a static property on the resulting component which is used as a screen (`MyScreen.navigationOptions = someoptions`). In React Navigation, screens specify `navigationOptions` as a static property on the component, so the behavior is consistent and this behaviour is expected.

> The most intuitive behavior is that, if something has a header, you should be able to set the headerStyle on it

This is how it works currently, the resulting `React.Component` from the `TabNavigator` has a header, so you specify a `navigationOptions` static property on it with `headerStyle` and it's respected.

> and, if it doesn't have a headerStyle, it inherits the headerStyle from its parent.

I'm not sure it's intuitive. When you don't specify `headerStyle` for a `StackNavigator`, you don't expect it to inherit `headerStyle` from an unrelated parent `StackNavigator`. Also I don't think you mean parent, you probably mean child, since in the example `TabNavigator` is a child screen of `StackNavigator`. And `StackNavigator` does respect the static `navigationOptions` specified on the screen.

> This works with other header properties, which emphasizes why the current behavior is not intuitive.

Can you post a repro? I couldn't find a repro in either of the issues which shows that other options work. If it does, it's a bug.

And if you want to change the behavior of how the options are resolved, please open a new issue with a proposal of how the navigation options could be resolved for nested navigators and we can discuss on how to implement it.

**Comment by llamaluvr** [comment]
@satya164 here's a snack demonstrating the OP's issue: https://snack.expo.io/@llamaluvr/react-navigation-nested-headerstyle-issue. The `headerStyle` on the `TabNavigator` doesn't actually matter here, so I removed it.

The expected behavior in the snack is that the left tab's header is yellow (because it is set on the child), and the right tab's header is blue (because it inherits from the parent). But the actual behavior is that both tabs are blue. Additionally, you'll see the same thing with the `headerLeft` option - it should take the child but actually takes the parent. If you remove the `headerStyle` or `headerLeft` props from the parent `StackNavigator`, then then child options come through (so apparently it actually isn't inconsistent with other options in this case).

As a control experiment, here is a `StackNavigator` with the screen directly under it- in this case, the screen `navigationOptions` do override the `StackNavigator`'s: https://snack.expo.io/@llamaluvr/react-navigation-nested-header-style-issue---only-stacknavigators

The broader goal here is to be able to make the header on one tab different from other headers in the rest of the stack without setting each header on each screen individually. The current behavior makes a pretty common scenario quite difficult and fragile to implement. In our case, we have styles that need to carry through the entire stack, but then we need to override them on tabs where we implement a search control that's contiguous with the header.

------

I'm having trouble reproducing the inconsistency I thought I saw with how different `navigationOptions` did or didn't override (e.g., `headerStyle` didn't override but `headerLeft` did). I realize where I tried this was a few versions ago, but more importantly, there was a custom navigator based on a `TabRouter` involved (for implementing a switching of routes based on a dropdown menu). So, more like Stack -> Tab -> Tab -> Screen than Stack -> Tab -> Screen. If I reproduce it, I'll be sure to post it. I appreciate you noting the difference between the constructor `navigationOptions` and the static prop - definitely gives me something else play around with there.

**Comment by chochihim** [comment]
Thanks @llamaluvr for explaining in details about my problem.

> If you remove the headerStyle or headerLeft props from the parent StackNavigator, then then child options come through (so apparently it actually isn't inconsistent with other options in this case)

I create a [snack](https://snack.expo.io/H1RDFDnQz) to demonstrate this.

Given this hierarchy Stack > Tab > Screen, the problem is navigation option (in this example, `headerStyle` or `headerLeft`) defined on Screen does not take effect if Stack has defined the same option. I think Screen's options should override Stack's.

**Comment by Rzygan** [comment]
april 2018, and the bug persists ...

**Comment by nmartin867** [comment]
I am experiencing the same issue but, unlike @chochihim, I am not setting navigationOptions on any of the parent navigators.

My application hierarchy is  Stack -> Tab -> Screen. navigationOptions is only set on 'HomeScreen' and they are ignored. The navigationOptions are applied, as expected, when HomeScreen is a direct child of the StackNavigator. 

Here is a snack: https://snack.expo.io/HkDsr55WQ

**Comment by quachsimon** [comment]
Has this issue been resolved? I am experiencing this on version `1.5.12` still with `headerLeft` where if I set it in my `StackNavigator` then my components `MyScreen.navigationOptions` is not working. If I do not include it in the stack then `MyScreen.navigationOptions` works when I have the header left set in there.

**Comment by brentvatne** [comment]
we don't support 1.5.12 anymore. i'm not aware of this issue in the 2.x series of releases, where the logic around resolving options changed: https://reactnavigation.org/docs/en/navigation-options-resolution.html

cc @nmartin867 as that's the solution your question too.

if anyone still has any problems please create a new issue and follow the template, thanks!
